src/sinks/influxdb/logs.rs

```diff
@@ -0,0 +1,657 @@
+    };
+}
+
+#[derive(Deserialize, Serialize, Debug, Eq, PartialEq, Clone, Derivative)]
+#[serde(rename_all = "snake_case")]
+#[derivative(Default)]
+pub enum Encoding {
+    #[derivative(Default)]
+    Default,
+}
+
+inventory::submit! {
+    SinkDescription::new_without_default::<InfluxDBLogsConfig>("influxdb_logs")
+}
+
+#[typetag::serde(name = "influxdb_logs")]
+impl SinkConfig for InfluxDBLogsConfig {
+    fn build(&self, cx: SinkContext) -> crate::Result<(super::RouterSink, super::Healthcheck)> {
+        let healthcheck = healthcheck(
+            self.clone().endpoint,
+            self.clone().influxdb1_settings,
+            self.clone().influxdb2_settings,
+            cx.resolver(),
+        )?;
+
+        let batch = self.batch.unwrap_or(bytesize::mib(1u64), 1);
+        let request = self.request.unwrap_with(&REQUEST_DEFAULTS);
+
+        let sink =
+            BatchedHttpSink::new(self.clone(), Buffer::new(false), request, batch, None, &cx)
+                .sink_map_err(|e| error!("Fatal influxdb_logs sink error: {}", e));
+
+        Ok((Box::new(sink), healthcheck))
+    }
+
+    fn input_type(&self) -> DataType {
+        DataType::Log
+    }
+
+    fn sink_type(&self) -> &'static str {
+        "influxdb_logs"
+    }
+}
+
+impl HttpSink for InfluxDBLogsConfig {
+    type Input = Vec<u8>;
+    type Output = Vec<u8>;
+
+    fn encode_event(&self, event: Event) -> Option<Self::Input> {
+        let mut output = String::new();
+        let mut event = event.into_log();
+
+        let tag_keys: Vec<String> = [
+            vec!["host".to_owned(), "source_type".to_owned()],
+            self.tags.clone(),
+        ]
+        .concat();
+
+        // Measurement
+        let measurement = encode_namespace(&self.namespace, &"vector");
+
+        // Timestamp
+        let timestamp =
+            if let Some(Value::Timestamp(ts)) = event.remove(log_schema().timestamp_key()) {
+                encode_timestamp(Some(ts))
+            } else {
+                encode_timestamp(None)
+            };
+
+        // Tags
+        let tags: BTreeMap<String, String> = event
+            .all_fields()
+            .filter(|(key, _)| tag_keys.contains(key))
+            .map(|(key, value)| (key, value.to_string_lossy()))
+            .collect();
+
+        // Fields
+        let fields: HashMap<String, Field> = event
+            .all_fields()
+            .filter(|(key, _)| !tag_keys.contains(key))
+            .map(|(key, value)| (key, value.to_field()))
+            .collect();
+
+        influx_line_protocol(
+            measurement,
+            "logs",
+            Some(tags),
+            Some(fields),
+            timestamp,
+            &mut output,
+        );
+
+        Some(output.into_bytes())
+    }
+
+    fn build_request(&self, events: Self::Output) -> http::Request<Vec<u8>> {
+        let settings = influxdb_settings(
+            self.influxdb1_settings.clone(),
+            self.influxdb2_settings.clone(),
+        )
```

Comment:
Can this be rewritten to avoid the `clone`? Either pass by reference or move the `settings` function into a structure containing the two settings values?

Reply:
What about move the `settings` into SinkConfig and initialise it in `build` method? It is suitable for you?

Reply:
Hoisting all possible initialization into the `build` method is the best solution, yes.

Reply:
Done